Build still fails when both -sJSPI and --emit-tsd are active

**Version of emscripten/emsdk:**
```bash
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.74 (1092ec30a3fb1d46b1782ff1b4db5094d3d06ae5)
clang version 20.0.0git (https:/github.com/llvm/llvm-project 322eb1a92e6d4266184060346616fa0dbe39e731)
Target: wasm32-unknown-emscripten
Thread model: posix
```

**Issue:**
*This is a follow-up to https://github.com/emscripten-core/emscripten/issues/23272 because it was closed.*

When enabling the **JSPI** flag and attempting to emit **Typescript definitions** (via --emit-tsd), the build fails. Since we disable Asyncify the `library_async` is not included and the linker complains about missing `emscripten_sleep()` function.

Demo to reproduce: https://github.com/mroohian/emscripten-jspi-demo

**Failing command line in full:**
```bash
# This command is executed by CMake
em++ -gsource-map \
     -lembind \
     --emit-tsd test1.d.ts \   # It works without this line
     -s EXPORT_NAME=Test1Module \
     -s JSPI \
     -s STRICT \
     -s USE_ES6_IMPORT_META \
     -s MODULARIZE \
     -s EXPORT_ES6=1 \
     --no-entry \
     test1.cpp.o \
     -o test1.js
```

**Full link command and output with `-v` appended:**
```shell
> emscripten-jspi-demo@1.0.0 cmake:build
> cd build.em && make

-- Configuring done (1.2s)
-- Generating done (0.0s)
-- Build files have been written to: /root/workspace/emscripten-jspi-demo/build.em
[ 33%] Building CXX object CMakeFiles/test1.dir/src/wasm/calc.cpp.o
[ 66%] Building CXX object CMakeFiles/test1.dir/src/wasm/test1.cpp.o
[100%] Linking CXX executable test1.js
em++: warning: -sASYNCIFY=2 (JSPI) is still experimental [-Wexperimental]
error: undefined symbol: emscripten_sleep (referenced by root reference (e.g. compiled C/C++ code))
warning: To disable errors for undefined symbols use `-sERROR_ON_UNDEFINED_SYMBOLS=0`
warning: _emscripten_sleep may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
Error: Aborting compilation due to previous errors
em++: error: '/root/workspace/emsdk/node/20.18.0_64bit/bin/node /root/workspace/emsdk/upstream/emscripten/src/compiler.mjs /tmp/tmpyi_gjmeb.json' failed (returned 1)
make[2]: *** [CMakeFiles/test1.dir/build.make:114: test1.js] Error 1
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/test1.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```

--- a/src/library_async.js
+++ b/src/library_async.js
@@ -38,6 +38,10 @@ addToLibrary({
     rewindArguments: {},
 #endif
     instrumentWasmImports(imports) {
+#if EMBIND_GEN_MODE
+      // Instrumenting is not needed when generating code.
+      return imports;
+#endif
 #if ASYNCIFY_DEBUG
       dbg('asyncify instrumenting imports');
 #endif
@@ -103,6 +107,10 @@ addToLibrary({
     },
 #endif
     instrumentWasmExports(exports) {
+#if EMBIND_GEN_MODE
+      // Instrumenting is not needed when generating code.
+      return exports;
+#endif
 #if ASYNCIFY_DEBUG
       dbg('asyncify instrumenting exports');
 #endif

--- a/src/settings_internal.js
+++ b/src/settings_internal.js
@@ -92,6 +92,10 @@ var EMBIND = false;
 // Whether a TypeScript definition file has been requested.
 var EMIT_TSD = false;
 
+// This will be true during the generation of code in run_embind_gen. Helpful
+// for detecting if either TSD file or embind AOT JS generation is running.
+var EMBIND_GEN_MODE = false;
+
 // Whether the main() function reads the argc/argv parameters.
 var MAIN_READS_PARAMS = true;
 

--- a/test/other/embind_tsgen_jspi.cpp
+++ b/test/other/embind_tsgen_jspi.cpp
@@ -1,8 +1,11 @@
 #include <emscripten/bind.h>
+#include <emscripten.h>
 
 using namespace emscripten;
 
-void sleep() {}
+void sleep() {
+  emscripten_sleep(0);
+}
 
 EMSCRIPTEN_BINDINGS(Test) {
   function("sleep", &sleep, async());

--- a/test/other/embind_tsgen_jspi.d.ts
+++ b/test/other/embind_tsgen_jspi.d.ts
@@ -1,17 +1,4 @@
 // TypeScript bindings for emscripten-generated code.  Automatically generated at compile time.
-declare namespace RuntimeExports {
-    let HEAPF32: any;
-    let HEAPF64: any;
-    let HEAP_DATA_VIEW: any;
-    let HEAP8: any;
-    let HEAPU8: any;
-    let HEAP16: any;
-    let HEAPU16: any;
-    let HEAP32: any;
-    let HEAPU32: any;
-    let HEAP64: any;
-    let HEAPU64: any;
-}
 interface WasmModule {
 }
 
@@ -19,4 +6,4 @@ interface EmbindModule {
   sleep(): Promise<void>;
 }
 
-export type MainModule = WasmModule & typeof RuntimeExports & EmbindModule;
+export type MainModule = WasmModule & EmbindModule;

--- a/test/test_other.py
+++ b/test/test_other.py
@@ -3530,7 +3530,7 @@ More info: https://emscripten.org
   @requires_jspi
   def test_embind_tsgen_jspi(self):
     self.run_process([EMXX, test_file('other/embind_tsgen_jspi.cpp'),
-                      '-lembind', '--emit-tsd', 'embind_tsgen_jspi.d.ts', '-sJSPI'] +
+                      '-lembind', '--emit-tsd', 'embind_tsgen_jspi.d.ts', '-sJSPI', '-sSTRICT', '--no-entry'] +
                      self.get_emcc_args())
     self.assertFileContents(test_file('other/embind_tsgen_jspi.d.ts'), read_file('embind_tsgen_jspi.d.ts'))
 

--- a/tools/link.py
+++ b/tools/link.py
@@ -1964,6 +1964,7 @@ def run_embind_gen(wasm_target, js_syms, extra_settings, linker_inputs):
   # Save settings so they can be restored after TS generation.
   original_settings = settings.backup()
   settings.attrs.update(extra_settings)
+  settings.EMBIND_GEN_MODE = True
 
   if settings.MAIN_MODULE and linker_inputs:
     # Copy libraries to the temp directory so they can be used when running
@@ -2001,11 +2002,6 @@ def run_embind_gen(wasm_target, js_syms, extra_settings, linker_inputs):
   setup_environment_settings()
   # Use a separate Wasm file so the JS does not need to be modified after emscripten.emscript.
   settings.SINGLE_FILE = False
-  if settings.ASYNCIFY == 2:
-    # JSPI is not needed to generate the definitions.
-    # TODO: when the emsdk node version supports JSPI, it probably makes more sense
-    # to enable it in node than disabling the setting here.
-    settings.ASYNCIFY = 0
   # Embind may be included multiple times, de-duplicate the list first.
   settings.JS_LIBRARIES = dedup_list(settings.JS_LIBRARIES)
   # Replace embind with the TypeScript generation version.

